File: python/sparktk/frame/ops/bin_column.py

def bin_column(self, column_name, bins=None, include_lowest=True, strict_binning=False, bin_column_name=None):
    """
    Summarize rows of data based on the value in a single column by sorting them
    into bins, or groups, based on a list of bin cutoff points or a specified number of
    equal-width bins.

    **Notes**

    #)  Bins IDs are 0-index, in other words, the lowest bin number is 0.
    #)  The first and last cutoffs are always included in the bins.
        When *include_lowest* is ``True``, the last bin includes both cutoffs.
        When *include_lowest* is ``False``, the first bin (bin 0) includes both
        cutoffs.

    :param column_name: Name of the column to bin
    :param bins: Either a single value representing the number of equal-width bins to create, or an array of values
     containing bin cutoff points. Array can be list or tuple. If an array is provided, values must be progressively
     increasing. All bin boundaries must be included, so, with N bins, you need N+1 values.
     Default is equal-width bins where the maximum number of bins is the Square-root choice
     :math:`\lfloor \sqrt{m} \rfloor`, where :math:`m` is the number of rows.
    :param include_lowest: Specify how the boundary conditions are handled. ``True`` indicates that the lower bound
     of the bin is inclusive. ``False`` indicates that the upper bound is inclusive. Default is ``True``.
    :param strict_binning: Specify how values outside of the cutoffs array should be binned. If set to ``True``, each
     value less than cutoffs[0] or greater than cutoffs[-1] will be assigned a bin value of -1. If set to ``False``,
     values less than cutoffs[0] will be included in the first bin while values greater than cutoffs[-1] will be
     included in the final bin.
    :param bin_column_name: The name for the new binned column.  Default is ``<column_name>_binned``
    :return: a list containing the edges of each bin

    Examples
    --------
    For these examples, we will use a frame with column *a* accessed by a Frame
    object *my_frame*:

    <hide>

    >>> frame = tc.to_frame([[1],[1],[2],[3],[5],[8],[13],[21],[34],[55],[89]], [('a', int)])
    -etc-

    </hide>
    >>> frame.inspect(n=11)
    [##]  a
    ========
    [0]    1
    [1]    1
    [2]    2
    [3]    3
    [4]    5
    [5]    8
    [6]   13
    [7]   21
    [8]   34
    [9]   55
    [10]  89

    Modify the frame with a column showing what bin the data is in, by
    specifying cutoffs for the bin edges.
    The data values should use strict_binning:

    >>> frame.bin_column('a', [5, 12, 25, 60], include_lowest=True,
    ... strict_binning=True, bin_column_name='binned_using_cutoffs')
    <progress>

    >>> frame.inspect(n=11)
    [##]  a   binned_using_cutoffs
    ==============================
    [0]    1                    -1
    [1]    1                    -1
    [2]    2                    -1
    [3]    3                    -1
    [4]    5                     0
    [5]    8                     0
    [6]   13                     1
    [7]   21                     1
    [8]   34                     2
    [9]   55                     2
    [10]  89                    -1

    <hide>
    >>> frame.drop_columns('binned_using_cutoffs')
    -etc-

    </hide>

    Modify the frame with a column showing what bin the data is in.
    The data value should not use strict_binning:

    >>> frame.bin_column('a', [5, 12, 25, 60], include_lowest=True,
    ... strict_binning=False, bin_column_name='binned_using_cutoffs')
    <progress>

    >>> frame.inspect(n=11)
    [##]  a   binned_using_cutoffs
    ==============================
    [0]    1                     0
    [1]    1                     0
    [2]    2                     0
    [3]    3                     0
    [4]    5                     0
    [5]    8                     0
    [6]   13                     1
    [7]   21                     1
    [8]   34                     2
    [9]   55                     2
    [10]  89                     2

    <hide>
    >>> frame.drop_columns('binned_using_cutoffs')
    -etc-

    </hide>
    Modify the frame with a column showing what bin the data is in.
    The bins should be lower inclusive:

    >>> frame.bin_column('a', [1,5,34,55,89], include_lowest=True,
    ... strict_binning=False, bin_column_name='binned_using_cutoffs')
    <progress>

    >>> frame.inspect( n=11 )
    [##]  a   binned_using_cutoffs
    ==============================
    [0]    1                     0
    [1]    1                     0
    [2]    2                     0
    [3]    3                     0
    [4]    5                     1
    [5]    8                     1
    [6]   13                     1
    [7]   21                     1
    [8]   34                     2
    [9]   55                     3
    [10]  89                     3

    <hide>
    >>> frame.drop_columns('binned_using_cutoffs')
    -etc-

    </hide>
    Modify the frame with a column showing what bin the data is in.
    The bins should be upper inclusive:

    >>> frame.bin_column('a', [1,5,34,55,89], include_lowest=False,
    ... strict_binning=True, bin_column_name='binned_using_cutoffs')
    <progress>

    >>> frame.inspect( n=11 )
    [##]  a   binned_using_cutoffs
    ==============================
    [0]    1                     0
    [1]    1                     0
    [2]    2                     0
    [3]    3                     0
    [4]    5                     0
    [5]    8                     1
    [6]   13                     1
    [7]   21                     1
    [8]   34                     1
    [9]   55                     2
    [10]  89                     3

    <hide>
    >>> frame.drop_columns('binned_using_cutoffs')
    -etc-

    </hide>
    Modify the frame with a column of 3 equal-width bins.  This also
    returns the cutoffs that were used for creating the bins.

    >>> cutoffs = frame.bin_column('a', 3, bin_column_name='equal_width_bins')

    >>> print cutoffs
    [1.0, 30.333333333333332, 59.666666666666664, 89.0]

    >>> frame.inspect(n=frame.count())
    [##]  a   equal_width_bins
    ==========================
    [0]    1                 0
    [1]    1                 0
    [2]    2                 0
    [3]    3                 0
    [4]    5                 0
    [5]    8                 0
    [6]   13                 0
    [7]   21                 0
    [8]   34                 1
    [9]   55                 1
    [10]  89                 2

    """
    if not isinstance(bins, list):
        bins = [bins]
    return list(self._scala.binColumn(column_name,
                                self._tc.jutils.convert.to_scala_option_list_double(bins),
                                include_lowest,
                                strict_binning,
                                self._tc.jutils.convert.to_scala_option(bin_column_name)))

File: core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/ops/binning/BinColumn.scala
package org.trustedanalytics.sparktk.frame.internal.ops.binning

import org.trustedanalytics.sparktk.frame.internal.{ FrameState, FrameTransformWithResult, BaseFrame, FrameTransformReturn }
import org.trustedanalytics.sparktk.frame.{ Column, DataTypes }

//todo - FrameBinColumn  (remove *Trait)
//todo - BinColumnTransform  (remove *Trait)
//todo - FrameBinColumnTranform  (remove *Trait)
trait BinColumnTransformWithResult extends BaseFrame {

  /**
   * Classify data into user-defined groups.
   *
   * Summarize rows of data based on the value in a single column by sorting them into bins, or groups, based on a
   * list of bin cutoff points.
   *
   * @note
   *       1. Bins IDs are 0-index, in other words, the lowest bin number is 0.
   *       1. The first and last cutoffs are always included in the bins. When ''include_lowest'' is ``True``, the last
   *       bin includes both cutoffs. When ''include_lowest'' is ``False``, the first bin (bin 0) includes both cutoffs.
   *
   * @param column the column to bin
   * @param bins If a single bin value is provided, it defines the number of equal-width bins that will be created.
   *             Otherwise, bins can be a sequence of bin edges. If a list of bin cutoff points is specified, they must
   *             be progressively increasing; all bin boundaries must be defined, so with N bins, N+1 values are required.
   *             If no bins are specified, the default is to create equal-width bins, where the default number of bins is
   *             the square-root of the number of rows.
   * @param includeLowest true means the lower bound is inclusive, where false means the upper bound is inclusive.
   * @param strictBinning if true, each value less than the first cutoff value or greater than the last cutoff value
   *                      will be assigned to a bin value of -1; if false, values less than the first cutoff value will
   *                      be placed in the first bin, and those beyond the last cutoff will go in the last bin
   * @param binColumnName The name of the new column may be optionally specified
   *
   */
  def binColumn(column: String,
                bins: Option[Seq[Double]],
                includeLowest: Boolean = true,
                strictBinning: Boolean = false,
                binColumnName: Option[String] = None): Seq[Double] = {

    execute(BinColumn(column, bins, includeLowest, strictBinning, binColumnName))
  }
}

case class BinColumn(column: String,
                     bins: Option[Seq[Double]],
                     includeLowest: Boolean,
                     strictBinning: Boolean,
                     binColumnName: Option[String]) extends FrameTransformWithResult[Seq[Double]] {

  override def work(state: FrameState): FrameTransformReturn[Seq[Double]] = {
    val columnIndex = state.schema.columnIndex(column)
    state.schema.requireColumnIsNumerical(column)
    val newColumnName = binColumnName.getOrElse(state.schema.getNewColumnName(column + "_binned"))

    val binnedRdd = bins match {
      case None | Some(Nil) =>
        DiscretizationFunctions.binEqualWidth(columnIndex, HistogramFunctions.getNumBins(None, state.rdd), state.rdd)
      case Some(List(n)) =>
        require(n.isValidInt, s"Number of equal-width bins must be a round number, but was ${n}.")
        DiscretizationFunctions.binEqualWidth(columnIndex, n.toInt, state.rdd)
      case Some(x) => DiscretizationFunctions.binColumns(columnIndex, x, includeLowest, strictBinning, state.rdd)
    }

    // Return frame state and cutoffs array
    FrameTransformReturn(FrameState(binnedRdd.rdd, state.schema.copy(columns = state.schema.columns :+ Column(newColumnName,
      DataTypes.int32))), binnedRdd.cutoffs)
  }

}

